README.md
# Json API
Provides the implementation to communicate with an [json api compatible](https://jsonapi.org/) endpoint.
The package is build with [axios](https://www.npmjs.com/package/axios) as http client and utilizes [jsonapi-serializer](https://www.npmjs.com/package/json-api-serializer) to normalize the results into javascript objectes.

## Installation
Install this package using npm:

```
npm install @labor-digital/json-api
```

## Usage
```typescript
import {JsonApiState} from "./lib/JsonApiState"; 
import {JsonApi} from "./lib/JsonApi";

// Create the instance
const api = new JsonApi({baseUrl: "https://example.org/api/resources"});

// Request the resources of type "myResource" that are on page 2
// The query object has some types already defined but as specified in the json-api 
// definition, it can be extended with your own keys
const results = api.get("myResource", {page: {number: 2}});

// Iterate all results in the list and print the value of "myKey" into the console
results.forEach((result: JsonApiState) => {
    console.log(result.get("myKey"));
});

// You can also request a single resource by using the unique id as a lookup query
const state = api.getSingle("myResource", 12);
console.log(state.get("myKey"));
```

## Documentation
The documentation / API can be found [here](https://json-api.labor.tools) but is also rendered at "./doc/index.html" for local usage

## Postcardware
You're free to use this package, but if it makes it to your production environment we highly appreciate you sending us a postcard from your hometown, mentioning which of our package(s) you are using.

Our address is: LABOR.digital - Fischtorplatz 21 - 55116 Mainz, Germany

We publish all received postcards on our [company website](https://labor.digital).

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [4.0.1](https://bitbucket.org/labor-digital/json-api/compare/v4.0.0...v4.0.1) (2021-06-07)

## [4.0.0](https://bitbucket.org/labor-digital/json-api/compare/v3.7.1...v4.0.0) (2021-04-06)


### ⚠ BREAKING CHANGES

* Update dependency of @labor-digital/helferlein to 4.0.5
or higher

### Features

* update dependencies ([b5f17c3](https://bitbucket.org/labor-digital/json-api/commit/b5f17c3d857f4bfe80f000cbdc2eff95472b1878))
* update dependencies ([c0ad2c5](https://bitbucket.org/labor-digital/json-api/commit/c0ad2c5d58f9a2f95a25781b9bb482885607f835))

### [3.7.1](https://bitbucket.org/labor-digital/json-api/compare/v3.7.0...v3.7.1) (2020-10-09)


### Bug Fixes

* **JsonApi:** resolve links and meta from single resources as well ([43a2f70](https://bitbucket.org/labor-digital/json-api/commit/43a2f7052445749978fb6dff104810aef3a64cbe))

## [3.7.0](https://bitbucket.org/labor-digital/json-api/compare/v3.6.3...v3.7.0) (2020-09-30)


### Features

* **Collection:** make collections iterable ([371ca76](https://bitbucket.org/labor-digital/json-api/commit/371ca76b2372b2e20a59a43401d2a247553089aa))
* **Resource:** add reactive getters for a resource object ([8d8945f](https://bitbucket.org/labor-digital/json-api/commit/8d8945fe25ccb4143c781568aca109bd5c15cbd9))


### Bug Fixes

* update dependencies ([d61c921](https://bitbucket.org/labor-digital/json-api/commit/d61c92113ee79f5a4d1eab555df7df713f27171d))

### [3.6.3](https://bitbucket.org/labor-digital/json-api/compare/v3.6.2...v3.6.3) (2020-07-20)


### Bug Fixes

* update dependencies ([fd6f79f](https://bitbucket.org/labor-digital/json-api/commit/fd6f79f36c3e69d7bdbb8e0c4473165edc8c699c))

### [3.6.2](https://bitbucket.org/labor-digital/json-api/compare/v3.6.1...v3.6.2) (2020-07-20)


### Bug Fixes

* update dependencies ([09f46f9](https://bitbucket.org/labor-digital/json-api/commit/09f46f978ac3ce2b03e7c129eb2545e1f961699b))

### [3.6.1](https://bitbucket.org/labor-digital/json-api/compare/v3.6.0...v3.6.1) (2020-06-19)


### Bug Fixes

* **JsonApi:** make sure different requests don't cancel each other out ([a884e76](https://bitbucket.org/labor-digital/json-api/commit/a884e76251e66e8a441aba6deec272b19878e5f3))

## [3.6.0](https://bitbucket.org/labor-digital/json-api/compare/v3.5.0...v3.6.0) (2020-05-22)


### Features

* deprecate JsonApiState + JsonApiStateList in favour of Resource and Collection ([0cd4e10](https://bitbucket.org/labor-digital/json-api/commit/0cd4e10baef3676fd8ce27ad5409427f7cfcfb6d))
* update dependencies ([36f3292](https://bitbucket.org/labor-digital/json-api/commit/36f3292b73e8241b2cdd35ebea8a0a91982b87a6))

## [3.5.0](https://bitbucket.org/labor-digital/json-api/compare/v3.4.2...v3.5.0) (2020-04-19)


### Features

* update dependencies ([9fd3c0f](https://bitbucket.org/labor-digital/json-api/commit/9fd3c0fe01d7884d81991856c9afe1f7657f8ec8))


### Bug Fixes

* **JsonApi:** make sure all 2xx response codes are valid not only 200 ([5571082](https://bitbucket.org/labor-digital/json-api/commit/5571082f657457e368ec25fa86fba0babbe144e9))

### [3.4.2](https://bitbucket.org/labor-digital/json-api/compare/v3.4.1...v3.4.2) (2020-03-23)

### [3.4.1](https://bitbucket.org/labor-digital/json-api/compare/v3.4.0...v3.4.1) (2020-03-23)

## 3.4.0 (2020-03-23)


### Features

* initial public release ([196ce3c](https://bitbucket.org/labor-digital/json-api/commit/196ce3cb9f0f906c7da595d9fc78a276b1cbcc1d))


### Bug Fixes

* fix missing dependencies in code after public release ([83518ad](https://bitbucket.org/labor-digital/json-api/commit/83518ad4eb2e947d5d9c0d335b321d4539c0d7ba))
* fix missing dependency after public release ([acbe0c3](https://bitbucket.org/labor-digital/json-api/commit/acbe0c388a34598778357ed426ab0821827f5ccf))

# [3.3.0] (2020-01-10)


### Features

* **JsonApi:** add server response information (status and headers) to the json api response object ([19e7c58])
* update dependencies ([77b65a9])



# [3.2.0] (2020-01-07)


### Features

* **State:** add new hasValue() method on the state object, to check for empty values ([ab2f56b])
* update dependencies ([6eb8527])



# [3.1.0] (2019-12-31)


### Features

* update dependencies ([e3bc00d])



# [3.0.0] (2019-12-30)


### Features

* remove translation helper ([e713caa])
* update dependencies ([c2d0d5a])


### BREAKING CHANGES

* translation helper removed + simplify file tree



## [2.0.2] (2019-12-29)



## [2.0.1] (2019-11-27)


### Bug Fixes

* remove jsonapi-normalizer dependency ([4736bb4])



# [2.0.0] (2019-11-27)


### Features

* simplify objects and use jsonapi-normalizer as de-serializer ([46dbf24])


### BREAKING CHANGES

* the deserialized objects may now look different than
before + changed public api of some objects



# [1.6.0] (2019-11-08)


### Features

* update dependencies ([53f99ac])



# [1.5.0] (2019-11-08)


### Features

* update dependencies ([17628a3])



# [1.4.0] (2019-10-25)


### Bug Fixes

* add error handling for JsonApi.getSingle() method when the json api response was inconsistent ([9f83ac9])


### Features

* update dependencies to latest version ([f5ab8e1])



# [1.3.0] (2019-10-15)


### Features

* better implementation for getSingle() ([eccd345])



# [1.2.0] (2019-09-26)


### Features

* add new features and better state handling ([2b2d261])



# 1.1.0 (2019-08-11)


### Bug Fixes

* remove doc generation from pipeline scripts ([7c8522a])
* remove doc generation from pipeline scripts ([8414fb6])
* remove tests from pipeline scripts ([7f3ecf6])


### Features

* initial commit ([fb1e57c])


src/State.ts
/*
 * Copyright 2019 LABOR.digital
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * Last modified: 2019.09.03 at 08:58
 */

import {getPath, isArray, isEmpty, isPlainObject, isString, isUndefined, PlainObject} from "@labor-digital/helferlein";

/**
 * A simple store to hold and manipulate a generic state
 */
export class State {
	
	/**
	 * The registered translations on this instance
	 */
	protected _state: PlainObject;
	
	/**
	 * Inject the initial state object
	 * @param initialState
	 */
	constructor(initialState: PlainObject | any) {
		if (!isPlainObject(initialState)) {
			if (!isEmpty(initialState)) console.error("Invalid state given!", initialState);
			this._state = {};
		} else this._state = initialState;
	}
	
	/**
	 * Returns the value of the initial state
	 * @param key
	 * @param fallback
	 */
	public get(key: string | Array<string>, fallback?) {
		if (isArray(key) || isString(key) && key.indexOf(".") !== -1) return getPath(this._state, key, fallback);
		if (isPlainObject(this._state) && !isUndefined(this._state[(key as string)])) return this._state[(key as string)];
		return fallback;
	}
	
	/**
	 * Returns true if a certain key exists in the state
	 * @param key
	 */
	public has(key: string | Array<string>): boolean {
		return !isUndefined(this.get(key));
	}
	
	/**
	 * Similar to "has" but also checks if the value on a key is "empty". If the value is considered empty this method will return false
	 * Empty values are: NULL, undefined, 0, "", " ", {}, [] and empty Maps and Sets
	 * @param key
	 * @param includeZero By default zero (0) is not seen as "empty" if you set this to true, it will be, tho
	 */
	public hasValue(key: string | Array<string>, includeZero?: boolean): boolean {
		const value = this.get(key);
		if (isUndefined(value)) return false;
		return !isEmpty(value, includeZero);
	}
	
	/**
	 * Returns all key and value pairs that are currently inside the state
	 */
	public getAll(): PlainObject {
		return {...this._state};
	}
	
	/**
	 * Destroys the internal state object
	 */
	public destroy() {
		this._state = undefined;
	}
}

src/JsonApiState.ts
/*
 * Copyright 2019 LABOR.digital
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * Last modified: 2019.11.25 at 20:30
 */

import {Resource} from "./Elements/Resource";

/**
 * @deprecated will be removed in v4.0 use "Resource" instead!
 */
export class JsonApiState extends Resource {
}

src/JsonApiStateList.ts
/*
 * Copyright 2019 LABOR.digital
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * Last modified: 2019.11.26 at 20:24
 */

import {Collection} from "./Elements/Collection";

/**
 * @deprecated will be removed in v4.0 use "Collection" instead!
 */
export class JsonApiStateList extends Collection {
}

src/Elements/Resource.ts
/*
 * Copyright 2020 LABOR.digital
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * Last modified: 2020.05.22 at 21:43
 */


import {forEach} from "@labor-digital/helferlein";
import {JsonApiElementInterface, JsonApiElementType, JsonApiResponse} from "../JsonApi.interfaces";
import {State} from "../State";

export class Resource extends State implements JsonApiElementInterface {
	
	/**
	 * Holds a reference to the response object
	 */
	protected _response: JsonApiResponse;
	
	/**
	 * JsonApiSingleState constructor
	 *
	 * @param response
	 */
	constructor(response: JsonApiResponse) {
		super(response.data);
		this._response = response;
		
		// Define reactive getters
		forEach(response.data, (v, k) => {
			Object.defineProperty(this, k, {
				get: function () {
					return v;
				}
			});
		});
	}
	
	/**
	 * Returns the type of state
	 */
	public get jsonElementType(): JsonApiElementType {
		return "item";
	}
	
	/**
	 * Returns the reference to the raw response object
	 */
	public get response(): JsonApiResponse {
		return this._response;
	}
}


src/JsonApi.interfaces.ts
/*
 * Copyright 2019 LABOR.digital
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * Last modified: 2019.08.11 at 21:27
 */

import {PlainObject} from "@labor-digital/helferlein";
import {AxiosInstance} from "axios";
import {Resource} from "./Elements/Resource";
import {JsonApiState} from "./JsonApiState";

export type JsonApiElementType = "item" | "collection";

export interface JsonApiElementInterface {
	/**
	 * Defines which type of element we work with
	 */
	jsonElementType: JsonApiElementType
}

export interface CollectionForEachCallbackType extends Function {
	/**
	 * Is called for every element of the iterated object
	 * @param value The current value
	 * @param key The current key
	 */
	(value?: Resource, key?: any): boolean | any;
}

export interface JsonApiArguments {
	
	/**
	 * If given, this will override all other arguments.
	 * If not given we create our own axios instance
	 */
	axios?: AxiosInstance
	
	/**
	 * The base url for the api.
	 * If not given we will assume "$host/api/resources"
	 */
	baseUrl?: string
}

export interface JsonApiGetQuery {
	[key: string]: any;
	
	/**
	 * The filter object for this query
	 */
	filter?: PlainObject,
	
	/**
	 * The array of sort fields.
	 * Prefix the field name with a "-" to sort descending
	 */
	sort?: Array<string> | string
	
	/**
	 * The list of included objects in this request
	 */
	include?: Array<string> | string
	
	/**
	 * The list of fields that should be requested
	 * should be an object representing the resource type and it's fields as arrays
	 */
	fields?: PlainObject;
	
	/**
	 * The pagination constraints
	 */
	page?: {
		[key: string]: any;
		number?: number,
		size?: number
	}
}

export interface JsonApiResponse {
	/**
	 * True if this is a single entity result
	 */
	isSingleResult: boolean;
	
	/**
	 * The http status code of the response
	 */
	status?: number;
	
	/**
	 * The list of http headers from the response
	 */
	headers?: PlainObject;
	
	/**
	 * Either a plain object representing a single result
	 * or an array representing multiple results in a response
	 */
	data: PlainObject | Array<PlainObject>;
	
	/**
	 * Can contain the link information given by the response
	 */
	links?: PlainObject;
	
	/**
	 * Additional meta information about this response
	 */
	meta?: {
		[key: string]: any;
		
		/**
		 * Optional pagination for this response
		 */
		pagination?: PlainObject
	}
}

export interface JsonApiPagination {
	/**
	 * The current page we are showing
	 * @var int
	 */
	page: Number;
	
	/**
	 * The number of all pages we have
	 * @var int
	 */
	pages: Number;
	
	/**
	 * The maximum number of items on a single page
	 * @var int
	 */
	pageSize: Number;
	
	/**
	 * The number of all items in the set
	 * @var int
	 */
	itemCount: Number;
}


/**
 * @deprecated will be removed in v4.0 use JsonApiPagination instead!
 */
export interface ApiPagination extends JsonApiPagination {
}

/**
 * @deprecated will be removed in v4.0 use CollectionForEachCallbackType instead!
 */
export interface StateListForEachCallbackType extends Function {
	/**
	 * Is called for every element of the iterated object
	 * @param value The current value
	 * @param key The current key
	 */
	(value?: JsonApiState, key?: any): boolean | any;
}

src/Elements/Collection.ts
/*
 * Copyright 2020 LABOR.digital
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * Last modified: 2020.05.22 at 21:45
 */

import {forEach, getPath, hasPath, isPlainObject, isUndefined, map, PlainObject} from "@labor-digital/helferlein";
import {
	CollectionForEachCallbackType,
	JsonApiElementInterface,
	JsonApiElementType,
	JsonApiPagination,
	JsonApiResponse
} from "../JsonApi.interfaces";
import {State} from "../State";
import {Resource} from "./Resource";

export class Collection implements JsonApiElementInterface, Iterable<Resource> {
	
	/**
	 * The pagination object or null
	 */
	protected _pagination: JsonApiPagination | null;
	
	/**
	 * Additional metadata that was passed by the api
	 */
	protected _meta: State;
	
	/**
	 * Holds a reference to the response object
	 */
	protected _response: JsonApiResponse;
	
	/**
	 * The list of states we are holding -> as array by their given order
	 */
	protected _byOrder: Array<Resource>;
	
	/**
	 * The list of states we are holding -> by their entity id
	 */
	protected _byId: PlainObject<Resource>;
	
	/**
	 * JsonApiSingleState constructor
	 *
	 * @param response
	 */
	constructor(response: JsonApiResponse) {
		this._response = response;
		
		// Create pagination reference
		if (hasPath(response, ["meta", "pagination"])) {
			const p = response.meta.pagination;
			this._pagination = {
				page: !isUndefined(p.currentPage) ? p.currentPage : p.current_page,
				pages: !isUndefined(p.totalPages) ? p.totalPages : p.total_pages,
				pageSize: !isUndefined(p.perPage) ? p.perPage : p.per_page,
				itemCount: p.count
			};
		}
		
		// Fetch meta from the response
		const meta = getPath(response, ["meta"], {});
		this._meta = new State(meta);
		
		// Build the list reference
		this._byOrder = [];
		this._byId = {};
		forEach(response.data, (val: PlainObject) => {
			const state = new Resource({
				isSingleResult: true,
				data: val,
				meta: response.meta,
				links: response.links
			});
			this._byOrder.push(state);
			this._byId[state.get("id")] = state;
		});
	}
	
	/**
	 * Returns the type of state
	 */
	public get jsonElementType(): JsonApiElementType {
		return "collection";
	}
	
	/**
	 * Returns the reference to the raw response object
	 */
	public get response(): JsonApiResponse {
		return this._response;
	}
	
	/**
	 * Iterates over all main entities in this state
	 * @param callback
	 */
	public forEach(callback: CollectionForEachCallbackType) {
		return forEach(this._byOrder, (item: Resource) => {
			return callback(item, item.get("id"));
		});
	}
	
	/**
	 * Returns a single item of the list by the entities id
	 * @param id
	 */
	public get(id: string | number): Resource | undefined {
		if (isUndefined(this._byId[id])) return undefined;
		return this._byId[id];
	}
	
	/**
	 * Returns true if a certain item with the given id exists in the state list
	 * @param id
	 */
	public has(id: string | number): boolean {
		return !isUndefined(this.get(id));
	}
	
	/**
	 * Returns the raw state array
	 */
	public getAll(): Array<Resource> {
		return this._byOrder;
	}
	
	/**
	 * Returns a raw array of all contained resources converted into a plain object
	 */
	public getRaw(): Array<PlainObject> {
		return map(this._byOrder, (v: Resource) => {
			return v.getAll();
		});
	}
	
	/**
	 * Returns true if there is a pagination, false if not
	 */
	public hasPagination(): boolean {
		return isPlainObject(this._pagination);
	}
	
	/**
	 * Returns the pagination for this api state
	 */
	public get pagination(): JsonApiPagination {
		return this._pagination;
	}
	
	/**
	 * Returns the meta information that was given by the api
	 */
	public get meta(): State {
		return this._meta;
	}
	
	/**
	 * Returns the iterator for the collection
	 */
	public [Symbol.iterator](): Iterator<Resource> {
		const that = this;
		return new class implements Iterator<Resource> {
			protected _i = 0;
			
			public next(...args: [] | [undefined]): IteratorResult<Resource, any> {
				if (this._i < that._byOrder.length) {
					return {value: that._byOrder[this._i++], done: false};
				} else {
					return {value: undefined, done: true};
				}
			}
		};
	}
}

src/JsonApi.ts
/*
 * Copyright 2019 LABOR.digital
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * Last modified: 2019.08.11 at 21:25
 */

import {
	debouncePromise,
	forEach,
	getGuid,
	getPath,
	isArray,
	isBool,
	isEmpty,
	isNull,
	isNumber,
	isObject,
	isPlainObject,
	isString,
	isUndefined,
	PlainObject
} from "@labor-digital/helferlein";
import axios, {AxiosInstance, AxiosResponse} from "axios";
import {Deserializer} from "jsonapi-serializer";
import {Collection} from "./Elements/Collection";
import {Resource} from "./Elements/Resource";
import {JsonApiArguments, JsonApiElementInterface, JsonApiGetQuery, JsonApiResponse} from "./JsonApi.interfaces";
import {JsonApiState} from "./JsonApiState";
import {JsonApiStateList} from "./JsonApiStateList";

export class JsonApi {
	
	/**
	 * A unique id for this api instance
	 */
	protected _guid;
	
	/**
	 * Holds the axios instance we use for our requests
	 */
	protected _axios: AxiosInstance;
	
	/**
	 * Constructor to inject the dependencies
	 * @param args
	 */
	constructor(args?: JsonApiArguments) {
		this._guid = getGuid("json-api-");
		if (isUndefined(args)) args = {};
		if (isUndefined(args.axios)) {
			if (!isString(args.baseUrl)) args.baseUrl = window.location.protocol + "//" + window.location.host + "/api/resources/";
			this._axios = axios.create({
				baseURL: args.baseUrl
			});
		} else {
			this._axios = args.axios;
		}
	}
	
	/**
	 * Returns the axios instance we use internally
	 */
	public get axios(): AxiosInstance {
		return this._axios;
	}
	
	/**
	 * Requests a resource collection from the server
	 *
	 * @param resourceType The resource type to request from the server
	 * @param query An optional query to use when requesting the resources
	 * @param debounceLimit Can be used to debounce multiple, subsequent requests for a certain number of milli-seconds,
	 *                      before they should be actually submitted to the server. Can be used to avoid multiple request
	 *                      for text inputs or similar occurrences.
	 */
	public getCollection(resourceType: string, query?: JsonApiGetQuery, debounceLimit?: number): Promise<Collection> {
		const queryString = this.makeQueryString(query);
		const guid = this._guid + "-" + resourceType + "-" + this.makeQueryString(query);
		debounceLimit = isNumber(debounceLimit) ? debounceLimit : 0;
		return debouncePromise(guid, () => {
			return this.axios
				.get("/" + resourceType + queryString)
				.then(res => this.handleAxiosResponse(res));
		}, debounceLimit, true);
	}
	
	/**
	 * Requests a single resource item from the server
	 *
	 * @param resourceType The resource type to request from the server
	 * @param id The id of the entity to request from the server. Set this to NULL if your endpoint does not
	 *             require a unique id for requesting a single entity
	 * @param query An optional query to use when requesting the resource
	 * @param debounceLimit Can be used to debounce multiple, subsequent requests for a certain number of milli-seconds,
	 *                      before they should be actually submitted to the server. Can be used to avoid multiple request
	 *                      for text inputs or similar occurrences.
	 */
	public getResource(resourceType: string, id: number | string | null, query?: JsonApiGetQuery, debounceLimit?: number): Promise<Resource> {
		const queryString = this.makeQueryString(query);
		const guid = this._guid + "-" + resourceType + "-" + id + "-" + this.makeQueryString(query);
		debounceLimit = isNumber(debounceLimit) ? debounceLimit : 0;
		return debouncePromise(guid, () => {
			return this.axios
				.get("/" + resourceType + (isNull(id) ? "" : "/" + id) + queryString)
				.then(res => this.handleAxiosResponse(res));
		}, debounceLimit, true);
	}
	
	/**
	 * Requests a resource collection from the server
	 *
	 * @param resourceType The resource type to request from the server
	 * @param query An optional query to use when requesting the resources
	 * @param debounceLimit Can be used to debounce multiple, subsequent requests for a certain number of milli-seconds,
	 *                      before they should be actually submitted to the server. Can be used to avoid multiple request
	 *                      for text inputs or similar occurrences.
	 *
	 * @deprecated will be removed in v4.0 - use getCollection() instead.
	 */
	public get(resourceType: string, query?: JsonApiGetQuery, debounceLimit?: number): Promise<JsonApiStateList> {
		return this.getCollection(resourceType, query, debounceLimit);
	}
	
	/**
	 * Requests a single resource entity from the server
	 *
	 * @param resourceType The resource type to request from the server
	 * @param id The id of the entity to request from the server. Set this to NULL if your endpoint does not
	 *             require a unique id for requesting a single entity
	 * @param query An optional query to use when requesting the resource
	 * @param debounceLimit Can be used to debounce multiple, subsequent requests for a certain number of milli-seconds,
	 *                      before they should be actually submitted to the server. Can be used to avoid multiple request
	 *                      for text inputs or similar occurrences
	 *
	 * @deprecated will be removed in v4.0 - use getResource() instead.
	 */
	public getSingle(resourceType: string, id: number | string | null, query?: JsonApiGetQuery, debounceLimit?: number): Promise<JsonApiState> {
		return this.getResource(resourceType, id, query, debounceLimit);
	}
	
	/**
	 * Can be used to format a query object into a url conform query string
	 * @param query
	 */
	public makeQueryString(query?: JsonApiGetQuery): string {
		const string = this.formatQuery(query);
		return string.length > 0 ? "?" + string : "";
	}
	
	/**
	 * Receives a raw json api response (presumably from an ajax request)
	 * and converts it either into a json api state, or a state list object, depending on the content
	 * @param response
	 * @deprecated Will be removed in v4.0 use makeResourceOrCollection() instead!
	 */
	public makeStateOrStateList(response: PlainObject | JsonApiResponse): Promise<JsonApiState | JsonApiStateList> {
		// Check if the response is already a json api response
		if (isBool(response.isSingleResult) && !isUndefined(response.data))
			return Promise.resolve(response.isSingleResult ?
				new JsonApiState(response as JsonApiResponse) :
				new JsonApiStateList(response as JsonApiResponse));
		
		// Create using deserialization
		return this.deserializeResponse(response).then(jsonResponse => {
			if (jsonResponse.isSingleResult)
				return new JsonApiState(jsonResponse);
			return new JsonApiStateList(jsonResponse);
		});
	}
	
	/**
	 * Receives a raw json api response (presumably from an ajax request)
	 * and converts it either into a json api state, or a state list object, depending on the content
	 * @param data The data to convert into either a resource or a collection
	 */
	public makeResourceOrCollection(data: PlainObject | JsonApiResponse | JsonApiElementInterface): Promise<Resource | Collection> {
		// Check if the data is already a json api element
		if (isString((data as JsonApiElementInterface).jsonElementType))
			return Promise.resolve(data as Collection);
		
		// Check if the response is already a json api response
		if (isBool((data as JsonApiResponse).isSingleResult) && !isUndefined((data as JsonApiResponse).data))
			return Promise.resolve((data as JsonApiResponse).isSingleResult ?
				new Resource(data as JsonApiResponse) :
				new Collection(data as JsonApiResponse));
		
		// Create using deserialization
		return this.deserializeResponse(data).then(jsonResponse => {
			if (jsonResponse.isSingleResult)
				return new Resource(jsonResponse);
			return new Collection(jsonResponse);
		});
	}
	
	/**
	 * Internal response handler to create
	 * @param response
	 */
	protected handleAxiosResponse(response: AxiosResponse): Promise<JsonApiState | JsonApiStateList> {
		// Throw an error if we got a non-200 response code
		if (response.status < 200 || response.status > 299) {
			let error = getPath(response.data, ["errors", "title"]);
			if (isUndefined(error)) error = response.statusText;
			return Promise.reject(new Error(error));
		}
		
		// Extend the response data by the status code and the response headers
		const data: JsonApiResponse = response.data;
		data.status = response.status;
		data.headers = response.headers;
		
		// Create a state object for the data
		return this.makeStateOrStateList(data);
	}
	
	/**
	 * Internal helper which formats the get query object into the correct query string
	 * @param query
	 * @param prefix
	 */
	protected formatQuery(query?: JsonApiGetQuery | PlainObject, prefix?: string): string {
		if (isUndefined(query)) return "";
		
		const output = [];
		
		const formatString = (key, value): string => {
			if (isObject(value)) return this.formatQuery(value, key);
			return encodeURIComponent(key) + "=" + encodeURIComponent(value);
		};
		
		forEach(query, (v, k) => {
			if (isEmpty(v)) v = "";
			if (!isUndefined(prefix)) k = prefix + "[" + k + "]";
			if (isArray(v)) v = v.join(",");
			const pair = formatString(k, v);
			if (pair === "") return;
			output.push(pair);
		});
		
		if (output.length === 0) return "";
		return output.join("&");
	}
	
	/**
	 * Internal helper which receives the raw result from axios and de-serializes
	 * the json api structure into a more speaking javascript object
	 * @param response
	 */
	protected deserializeResponse(response: PlainObject): Promise<JsonApiResponse> {
		// Fallback if we don't received a valid response
		if (isEmpty(response)) response = {data: {}};
		
		// Check if we got a data attribute
		if (isUndefined(response.data))
			throw new Error("Invalid json api response. There has to be a \"data\" node at the root level");
		
		// Deserialize the response
		return (new Deserializer({
			keyForAttribute: "camelCase",
			included: true
		})).deserialize(response).then((data) => {
			// Build our enriched output format
			const apiResponse: JsonApiResponse = {
				isSingleResult: !isArray(data),
				data: data
			};
			if (isPlainObject(response.meta ?? null)) apiResponse.meta = response.meta;
			if (isPlainObject(response.data.meta ?? null)) apiResponse.meta = response.data.meta;
			if (isPlainObject(response.links ?? null)) apiResponse.links = response.links;
			if (isPlainObject(response.data.links ?? null)) apiResponse.links = response.data.links;
			apiResponse.status = getPath(response, "status", 200);
			apiResponse.headers = getPath(response, "headers", {});
			
			// Done
			return apiResponse;
		});
	}
}